imageio/imageio-webp/src/main/java/com/twelvemonkeys/imageio/plugins/webp/LSBBitReader.java

```diff
@@ -11,46 +11,124 @@ public final class LSBBitReader {
     // TODO: Consider creating an ImageInputStream wrapper with the WebP implementation of readBit(s)?
 
     private final ImageInputStream imageInput;
-    int bitOffset = 0;
+    private int bitOffset = 64;
+    private long streamPosition = -1;
 
-    public LSBBitReader(ImageInputStream imageInput) {
+    /**
+     * Pre buffers up to the next 8 Bytes in input.
+     * Contains valid bits in bits 63 to {@code bitOffset} (inclusive).
+     * Should always be refilled to have at least 56 valid bits (if possible)
+     */
+    private long buffer;
+
+    public LSBBitReader(ImageInputStream imageInput){
         this.imageInput = imageInput;
     }
 
-    // TODO: Optimize this... Read many bits at once!
-    public long readBits(int bits) throws IOException {
-        long result = 0;
-        for (int i = 0; i < bits; i++) {
-            result |= (long) readBit() << i;
-        }
+    /**
+     * Reads the specified number of bits from the stream in an LSB-first way and advances the bitOffset.
+     * The underlying ImageInputStream will be advanced to the first not (completely) read byte.
+     * Requesting more than 64 bits will advance the reader by the correct amount and return the lowest 64 bits of
+     * the read number
+     *
+     * @param bits the number of bits to read
+     * @return a signed long built from the requested bits (truncated to the low 64 bits)
+     * @throws IOException if an I/O error occurs
+     * @see LSBBitReader#peekBits
+     */
+    public long readBits(int bits) throws IOException{
+        return readBits(bits, false);
+    }
 
-        return result;
+    /**
+     * Reads the specified number of bits from the buffer in an LSB-first way.
+     * Does not advance the bitOffset or the underlying input stream.
+     * As only 56 bits are buffered (in the worst case) peeking more is not possible without advancing the reader and
+     * as such disallowed.
+     *
+     * @param bits the number of bits to peek (max 56)
+     * @return a signed long built from the requested bits
+     * @throws IOException if an I/O error occurs
+     * @see LSBBitReader#readBits
+     */
+    public long peekBits(int bits) throws IOException{
+        if(bits > 56){
+            throw new IllegalArgumentException("LSBBitReader#peekBits: Tried peeking over 56");
```

Comment:
```suggestion
            throw new IllegalArgumentException("Tried peeking over 56");
```

(Class/method/line will be available from the stack trace)

We